### ISSUE TITLE
generated_by can't pull identity from Activity

### ISSUE THREAD

**Issue by jakebeal** [SynBioDex/pySBOL3#183]
Unlike some other compositional properties, the generated_by property doesn't know to pull an identity from an Activity that's added to it; you have to give it the identity property yourself

Minimal replication:

```
import sbol3

doc = sbol3.Document()
sbol3.set_namespace('https://example.org')

execution = sbol3.Activity('protocol_execution')
doc.add(execution)

experiment = sbol3.Experiment('results')
experiment.generated_by.append(execution)
```

**Comment by tcmitchell** [comment]
That is correct. The `generated_by` property is typed as a URI in the SBOL 3 specification, so the pySBOL3 module expects the programmer to supply a URI. `generated_by` is not a reference, which knows how to pull an identity from an object.

**Comment by jakebeal** [comment]
We've got two types of URI references, however, some that are just a URI and some where the URI is expected to be a reference to a certain class of object.  I believe that "lookup()" is sensitive to this distinction already.

Can't we have it that anything that we want to use "lookup()" on should also be able to accept an object and strip its URI out?

**Comment by tcmitchell** [comment]
If all you want is to have the URI property automatically extract the identity from any `Identified` and use it, we can do that. But my sense is that you think `generated_by` should work like the other referenced object properties like `Component.sequence` and `ComponentReference.feature` and it's just not going to work that way.

> We've got two types of URI references, however, some that are just a URI and some where the URI is expected to be a reference to a certain class of object. I believe that "lookup()" is sensitive to this distinction already.

Look at Figure 4 of the SBOL 3 specification (either 3.0 or draft 3.0.1). `generated_by` is a property that appears within the box depicting its class (`Identified`, Figure 5). The `lookup()` method is used for, in the language of the specification, shared aggregation association properties. These are depicted in Figure 4 with an arrow with an open diamond ("hasReferencedObject").

The `lookup()` method does not work with URI properties. It does work with referenced objects (shared aggregation association properties). The `lookup()` method is intended to find objects that are expected to be in the same `Document`. There is nothing about the URI properties that requires or even suggests that they might be in the same `Document`. Take, for example, `Component.type`, which is also a URI property (Figure 8). Would you expect to be able to `lookup()` the component type `https://identifiers.org/SBO:0000251` (SBO_DNA)?

> Can't we have it that anything that we want to use "lookup()" on should also be able to accept an object and strip its URI out?

You could certainly use `Document.find()` to look up a URI in a `Document`, which is exactly what the referenced object method `lookup()` does. But the URI properties do not have a `lookup()` method of their own the way referenced objects (shared aggregation association properties) have a `lookup()` method.

**Comment by bbartley** [comment]
Sounds like there is an issue with the specification rather than the library, then.  I think `generated_by` should indeed be a reference property rather than a URI property

**Comment by jakebeal** [comment]
You make a very good point --- I see that the spec is contradictory on this point. The diagram clearly indicates that it's a non-association property, but then the text below indicates that it is an association property with `prov:Activity`. Figure 27 in the prov appendix also shows it as an association property. A similar problem applies to `prov:wasDerivedFrom`.

This is a genuine specification problem, and I've opened an issue for it: https://github.com/SynBioDex/SBOL-specification/issues/435
The resolution of that issue will guide the resolution of this one.

**Comment by jakebeal** [comment]
This was resolved in the spec: now we need to check if the library is tracking the spec.

**Comment by tcmitchell** [comment]
This was fixed in commit https://github.com/SynBioDex/pySBOL3/pull/226/commits/adff16265d136611770026faf08abd6a2ca30a98 which was included in PR #226.
